File: 2019Fall-ML_DemoCode/Topic02_kNN/knnBank.r
# This file works on data available at a well-know repository of data
# hosted by the University of California-Irvine. The website for this
# dataset:
#  http://archive.ics.uci.edu/ml/datasets/Credit+Approval
# 
# The data contains information about individuals and whether they
# were approved  for credit. The problem is that the name of variables
# have been changed for privacy concerns. So we add  our own names.

# The data does not have any headers, so we add generic headers A1 to
# A16.

cat("This file compares many different classification techniques on a
credit approval data set from UCI web site.

We are downloading the data from a web site
Hit Enter to continue\n")
readline()

# This data has some parts missing (indicated by '?'). To tell the data
# frame that this we add na.string="?". Also note that the data is read
# off a web site directly, and not from a local file. Since there
# are no headers, we add headers A1 to A16 using sapply, and paste
# functions
bankData1 <-
    read.csv("http://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data",   na.strings="?",
        col.names=sapply(1:16, function(x){return (paste("A",x,sep=""))}))
cat("\nnorw(bankData) ", nrow(bankData1), "\n")

#Remove rows with <NA> in A2 column. 
bankData <- bankData1[!is.na(bankData1$A2),]# & !is.na(bankData1$A2)&!is.na(bankData1$A16),]
#(You can add A3 and A16 as well)
#bankData <- bankData1[!is.na(bankData1$A2)&!is.na(bankData1$A3)&!is.na(bankData1$A16),]
cat("Top rows of bankData:\n")
print(head(bankData))

cat("Bottom rows of bankData:\n")
print(tail(bankData))
readline()


# First, look at the connection between acceptance and A2:
cat("\n Hit Enter to see box plot A2 ~ acceptance:\n")
readline()
#dev.off()

boxplot(A2~A16, data=bankData, xlab="acceptance", ylab="A2")

# Now, look at the connection between acceptance and A3:
cat("\n Hit Enter to see box plot A3 ~ A16:\n")
readline()

#dev.set(dev.next())
boxplot(A3~A16, data=bankData, xlab="acceptance", ylab="A3")
readline()

#dev.off()

# We can now plot the combinations of A2 and A3 for acceptance (in
# blue) and no acceptance (in red)
cat("\n Hit Enter to see the scatterplot of A2 vs A3 for
acceptance cases:\n")
readline()
# Set up the graphics pane:

cat("\nSetting up the graphics pane:\n")
plot(bankData[,"A2"], bankData[,"A3"], col="white", 
    main="Scatter plot of bankData, A2 and A3 (orange reject bankData blue accept)", 
       xlab="A2", ylab="A3")
#readline()
points(bankData[bankData$A16=="+","A2"],
		bankData[bankData$A16=="+", "A3"], col="cornflowerblue",pch=18)
#cat("\nHit Enter to add the non-acceptance points:\n")
#readline()
points(bankData[bankData$A16=="-","A2"],
		bankData[bankData$A16=="-", "A3"], col="orange",pch=20)
####################Now build the knn model#############################

# There are several R packages that implement the kNN model. Two common
# ones are the klaR library (with the sknn function) and the kknn library (with
# the kknn function). The following code uses the kknn package. To use
# klaR, uncomment lines containing sknn, and comment out lines containing kknn.
# Also load the klaR library.

#cat("\nNow Running the knn model, loading the klaR library\n")
cat("\nNow Running the knn model, loading the kknn library\n")
#cat("\nHit Enter to start kNN:\n")
#readline()
# Two of many libraries providing the kNN algorithm are the klaR and the kknn
# libraries

#library("klaR")
library("kknn")

# We now create a set of points to test the model These new set of points form
# a fine grid in the A2-A3 plane. We collect this data into our test data frame
# called newPts0.
# 
# We next let the user choose $k$ and run the kNN algorithm for that k
# on the bankData data. We test the model on newPts0 data frame to see if the new
# points are classified as accepted for bankData or rejected. Finally we plot the
# newPts0 data juxtaposed on the training data. Those predicted to be accepted
# are colored blue, and those predicted to be rejected are colored red. 



newA2<-seq(mA2<-floor(min(bankData$A2)),MA2<-ceiling(max(bankData$A2)),by=1)
newA3<-seq(mA3<-floor(min(bankData$A3)),MA3<-ceiling(max(bankData$A3)),by=0.5)

# Creating a grid of points to find their classification under kNN
#newPts0<-data.frame(A2=c(), A3=c())
#for (i in newA2){
#   for (j in newA3){
#      newPts0<-rbind(data.frame(A2=i, A3=j), newPts0)
#   }
#}
newPts0<-data.frame(A2=kronecker(newA2, rep(1,length(newA3))), 
                    A3=rep(newA3,length(newA2)))
# Keep asking the user for k and build a corresponding kNN model:
repeat{
   cat("\nEnter the value k to be used for kNN (0 to exit):\n") 
   k<-as.numeric(readline())
   if(k<1) break
   cat("\nRunning kNN for k=",k,"\n")
   cat("\nHit Enter to see the results:\n")
   #knnModel <- sknn(A16 ~ A3 + A2, k=k, data=bankData)
   #knnPred<-predict(knnModel, newPts0, type="response")
   #newPts<-cbind(newPts0,data.frame(A16=knnPred$class))
   knnModel <- kknn(A16 ~ A3 + A2, k=k,train=bankData,
                   test=newPts0,kernel="rectangular")
   knnPred<-predict(knnModel, data=newPts0)
   newPts<-cbind(newPts0,data.frame(A16=knnPred))

   # Preparing for contour plots. The z matrix will contain predicted
   # probabilities for '-' (credit rejected):

   z=matrix(knnModel$prob,nrow=length(newA2),
             ncol = length(newA3),byrow = T)
   # Setting up the graphics pane. NewPts is a grid of points on the A2-A3
   # plane and then the knnModel is used to predict (red for predicting to reject 
   # credit, blue for predicting accept credit) We use image and contour to
   # draw the boundary between red and blue regions.  Image makes a bit more
   # resolution and instead of just blue and red, it changes color for larger
   # or smaller probabilities. This is sometimes called "heat map"
   
   #image(newA2,newA3,z,col=rainbow(20,start=.67, end=.09),main=paste(k,"NN"))
   image(newA2,newA3,z,col=rainbow(20,start=.55, end=.1),main=paste(k,"NN"))
   # For contour we set levels=0.5 so only the boundary between classes are
   # drawn, add = TRUE so that the contour is added to the existing graph

   contour(newA2,newA3,z,levels=c(0.5),add=TRUE,col="darkgreen", 
            drawlabels=FALSE,lwd=2)
   points(newPts[newPts$A16=='+',"A2"],newPts[newPts$A16=='+',"A3"],
					pch='.',cex=2,col="cornflowerblue")
   points(newPts[newPts$A16=='-',"A2"],newPts[newPts$A16=='-',"A3"],
				pch='.',cex=2,col="orange")
   dev.new()
   # Redraw contour without the image heatmap
   contour(newA2,newA3,z,levels=c(0.5),col="darkgreen",
        drawlabels=FALSE,lwd=2,main=paste(k,"NN"))
   points(newPts[newPts$A16=='+',"A2"],newPts[newPts$A16=='+',"A3"],
					pch='.',cex=2,col="cornflowerblue")
   points(newPts[newPts$A16=='-',"A2"],newPts[newPts$A16=='-',"A3"],
				pch='.',cex=2,col="orange")
   points(bankData[bankData$A16=="+","A2"],
		bankData[bankData$A16=="+", "A3"], col="cornflowerblue",pch=18)
   points(bankData[bankData$A16=="-","A2"],
		bankData[bankData$A16=="-", "A3"], col="orange",pch=20)
   # set up the confusion matrix:
   knn1Model=kknn(formula = A16 ~ A3 + A2, train = bankData, test = bankData, 
           k=k, kernel="rectangular")
   C= table(bankData$A16, predict(knn1Model))
   print(C)
   print(paste("Error rate:  ", (C[1,2]+C[2,1])/sum(C)))
}

File: 2019Fall-ML_DemoCode/Topic02_kNN/knnBank.py
# -*- coding: utf-8 -*-
"""
Created on Wed Aug 21 21:11:46 2019

@author: farid alizadeh
"""

# In this file we use the a cresdit approval data and two numerical featuers
# and run the knn algorithm. The value of k is determined by the user. 
# You can try different values and see the effect on error rate
#
# This file works on data available at a well-know repository of data
# hosted by the University of California-Irvine. The website for this
# dataset:
#  http://archive.ics.uci.edu/ml/datasets/Credit+Approval
# 
# The data contains information about individuals and whether they
# were approved  for credit. The problem is that the name of variables
# have been changed for privacy concerns. So we add  our own names.



#Libraries needed are sklearn's KNeighborClassifier numpy pandas matplot and 
#seaborn (for nicer plotting)
from sklearn.neighbors import KNeighborsClassifier
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

#In Python data can be downloaded directly from a website
url="http://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data"

#In this particular data set missing values are given by a '?' symbol (see the
#data documentation)
bankData=pd.read_csv(url,na_values="?",header=None)
#Make up column titles A1-A16. A16 is the target value: '+' means approved
#and '-' rejected for credit
bankData.columns=list(map(lambda x: 'A'+str(x+1),range(len(bankData.columns))))

#Remove all rows with a null (missing) vlaue
bankData=bankData.dropna()
print(bankData.head())
input('Hit Enter to continue')

# load the splitting code
runfile('splitTrainTest.py')

#split (no need for this script)
#trainDat,testDat=split_train_test(bankData, 0.2)

print('box plots showing dependence of credit approval to A2 and A3 variables:')

bankData.boxplot(column=['A2','A3'],by='A16')
plt.show()

print('Distribution of accpeted (blue) and rejected (orange) with A2 and A3 features:')

pd.Categorical(bankData['A16'],categories=['+','-'])

sb.relplot(x="A2", y="A3", hue="A16", data=bankData)
plt.show()

input('hit enter when done')

#Generate a grid of new data in XY plane and predict:
N=50
x_new=np.linspace(min(bankData.A2),max(bankData.A2),N)
y_new=np.linspace(min(bankData.A3),max(bankData.A3),N)
newPts=pd.DataFrame(columns=(['A2','A3']))

# The following double loop is natural, but for a script language
# like Python  it is very slow. I will shortly replace a 'vectorized'
# version 
for i in x_new:
    for j in y_new:
        newPts=newPts.append(pd.DataFrame([[i,j]],columns=['A2','A3']), ignore_index=True)
 
       
#newPts=newPts.reset_index()


####################Now build the knn model#############################

# The relevant library for knn classification:
from sklearn.neighbors import KNeighborsClassifier

# The user is expected enter a value of k and then we build 
# a knn model

k=int(input('Enter k (0 to end)'))

while (k>0):
   classifier = KNeighborsClassifier(n_neighbors=k)
   X=bankData.iloc[:,[1,2]] #extract columns 1&2 for 'A2' and 'A3'
   classifier.fit(X, bankData.A16) #build the model
   
   yPred=classifier.predict(newPts) #use the model to predict class
   yPredP=classifier.predict_proba(newPts) #and to predict probability
   newPts["A16"]=yPred #Add the predictions to the newPts data
   newPts["A16Prob"]=yPredP[:,0]

   print("plotting new points")
   sb.set()
   plt.scatter(newPts.A2[yPred=='+'],newPts.A3[yPred=='+'],c='cornflowerblue', marker='.')
   plt.scatter(newPts.A2[yPred=='-'],newPts.A3[yPred=='-'],c='orange', marker='.')
   plt.scatter(bankData.A2[bankData.A16=="+"], bankData.A3[bankData.A16=="+"], c='cornflowerblue')
   plt.scatter(bankData.A2[bankData.A16=="-"], bankData.A3[bankData.A16=="-"], c='orange',marker='.')
   #Uncomment to use seaborn software instead
   #sb.relplot(x="A2",y="A3", hue='A16', hue_order=['+','-'],marker='.', data=newPts)
   #sb.scatterplot(x=bankData[bankData.A16=='-'].A2,y=bankData[bankData=='-'].A3,c='orange')
   #sb.scatterplot(x=bankData[bankData.A16=='+'].A2,y=bankData[bankData=='+'].A3,c='cornflowerblue')
   
   xx,yy=np.meshgrid(x_new,y_new)
   zz=yPredP[:,0].reshape(len(xx),len(yy),order='F')#order F means column by column
   plt.contour(xx,yy,zz,levels=[0.5],colors='black')#contour for when prob=1/2
   plt.show()
   input('hit Enter to see the heat map')
  
   plt.contour(xx,yy,zz,levels=[0.5],colors='black')
   # For the heatmap we need to read the rows from bottom to top
   plt.imshow(zz[range(zz.shape[0]-1,0,-1),:],cmap="Spectral",extent=(min(x_new),max(x_new),min(y_new),max(y_new)))
   plt.colorbar()
   
   plt.show()

   
   from sklearn.metrics import classification_report, confusion_matrix
   yTrainPred=classifier.predict(bankData.iloc[:,[1,2]])
   yTrainPredP=classifier.predict_proba(bankData.iloc[:,[1,2]])
   C=confusion_matrix(bankData.A16, yTrainPred)
   s='Error rate for k={0:d} is {1:.3f}'
   print(s.format(k, 1-sum(C.diagonal())/C.sum()))

   print('confusion matrix:')
   print(C)
   print(classification_report(bankData.A16, yTrainPred))
   k=int(input("enter k, (0 to end):"))
   if k != 0:
       newPts=newPts.drop(["A16","A16Prob"],axis=1)
#end of while



